File: samples/Sample_S100/VB.NET_SBXBPCSample/frmAccessTz.vb
Option Strict Off
Option Explicit On
Imports System.Runtime.InteropServices

Public Class frmAccessTz
    Public Const DB_ACCESS_TIMEZONE_COUNT As Integer = 50
    Public Const DB_TIMESECTION_COUNT As Integer = 8
    Private Const DB_ALLCOUNT As Integer = DB_ACCESS_TIMEZONE_COUNT * DB_TIMESECTION_COUNT
    Dim timeZoneInfo(DB_ALLCOUNT * 4 - 1) As UInt32

    Private Sub frmAccessTz_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        timeZoneInit()
        DrawTimezoneInfo()
        dtStart.ShowUpDown = True
        dtEnd.ShowUpDown = True
    End Sub

    Private Sub timeZoneInit()
        Dim i As Integer, j As Integer, index As Integer
        For i = 0 To DB_ACCESS_TIMEZONE_COUNT - 1
            For j = 0 To DB_TIMESECTION_COUNT - 1
                index = i * DB_TIMESECTION_COUNT + j
                timeZoneInfo(index * 4 + 0) = 0
                timeZoneInfo(index * 4 + 1) = 0
                timeZoneInfo(index * 4 + 2) = 23
                timeZoneInfo(index * 4 + 3) = 59
            Next j
        Next i
    End Sub

    Private Sub DrawTimezoneInfo()
        Dim startHour As Integer, startMinute As Integer
        Dim endHour As Integer, endMinute As Integer
        Dim index As Integer
        Dim i As Integer, j As Integer
        Dim itemString As String

        lstTimeZone.Items.Clear()

        For i = 0 To DB_ACCESS_TIMEZONE_COUNT - 1
            For j = 0 To DB_TIMESECTION_COUNT - 1
                index = (i * DB_TIMESECTION_COUNT + j) * 4
                startHour = timeZoneInfo(index)
                startMinute = timeZoneInfo(index + 1)
                endHour = timeZoneInfo(index + 2)
                endMinute = timeZoneInfo(index + 3)
                itemString = "[Tz.]" & String.Format("{0:D2}-{1:D1} ", i, j)
                itemString += "[S]" & String.Format("{0:D2}:{1:D2} ", startHour, startMinute)
                itemString += "[E]" & String.Format("{0:D2}:{1:D2}", endHour, endMinute) + " "

                lstTimeZone.Items.Add(itemString)
            Next j
        Next i
    End Sub

    Private Sub lstTimeZone_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles lstTimeZone.SelectedIndexChanged
        If lstTimeZone.SelectedIndex = -1 Then
            Return
        End If

        Dim index As Integer = lstTimeZone.SelectedIndex
        dtStart.Value = New DateTime(2000, 1, 1, timeZoneInfo(index * 4), timeZoneInfo(index * 4 + 1), 0)
        dtEnd.Value = New DateTime(2000, 1, 1, timeZoneInfo(index * 4 + 2), timeZoneInfo(index * 4 + 3), 0)
    End Sub


    Private Sub cmdUpdate_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdUpdate.Click
        If lstTimeZone.SelectedIndex = -1 Then
            Return
        End If

        Dim index As Integer = lstTimeZone.SelectedIndex
        timeZoneInfo(index * 4 + 0) = dtStart.Value.Hour
        timeZoneInfo(index * 4 + 1) = dtStart.Value.Minute
        timeZoneInfo(index * 4 + 2) = dtEnd.Value.Hour
        timeZoneInfo(index * 4 + 3) = dtEnd.Value.Minute
        DrawTimezoneInfo()
    End Sub

    Private Sub frmAccessTz_FormClosing(ByVal sender As System.Object, ByVal e As System.Windows.Forms.FormClosingEventArgs) Handles MyBase.FormClosing
        frmMain.Visible = True
    End Sub

    Private Sub cmdExit_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdExit.Click
        Close()
    End Sub

    Private Sub cmdRead_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdRead.Click
        Dim bRet As Boolean
        Dim vErrorCode As Integer

        lblMessage.Text = "Working..."
        Application.DoEvents()

        bRet = frmMain.SBXPC1.EnableDevice(frmMain.gMachineNumber, False)
        If Not bRet Then
            lblMessage.Text = gstrNoDevice
            Exit Sub
        End If

        Dim gh As GCHandle = GCHandle.Alloc(timeZoneInfo, GCHandleType.Pinned)
        Dim AddrOftimZoneInfo As IntPtr = gh.AddrOfPinnedObject()
        bRet = frmMain.SBXPC1.GetDeviceLongInfo(frmMain.gMachineNumber, 3, AddrOftimZoneInfo.ToInt32())

        If bRet Then
            lblMessage.Text = "Success"
        Else
            frmMain.SBXPC1.GetLastError(vErrorCode)
            lblMessage.Text = ErrorPrint(vErrorCode)
        End If

        bRet = frmMain.SBXPC1.EnableDevice(frmMain.gMachineNumber, True)
        DrawTimezoneInfo()
    End Sub

    Private Sub cmdWrite_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdWrite.Click
        Dim bRet As Boolean
        Dim vErrorCode As Integer

        lblMessage.Text = "Working..."
        Application.DoEvents()

        bRet = frmMain.SBXPC1.EnableDevice(frmMain.gMachineNumber, False)
        If Not bRet Then
            lblMessage.Text = gstrNoDevice
            Exit Sub
        End If

        Dim gh As GCHandle = GCHandle.Alloc(timeZoneInfo, GCHandleType.Pinned)
        Dim AddrOftimZoneInfo As IntPtr = gh.AddrOfPinnedObject()
        bRet = frmMain.SBXPC1.SetDeviceLongInfo(frmMain.gMachineNumber, 3, AddrOftimZoneInfo.ToInt32())

        If bRet Then
            lblMessage.Text = "Success"
        Else
            frmMain.SBXPC1.GetLastError(vErrorCode)
            lblMessage.Text = ErrorPrint(vErrorCode)
        End If

        bRet = frmMain.SBXPC1.EnableDevice(frmMain.gMachineNumber, True)
    End Sub
End Class

File: sdk_extracted/20211204-SBXPC-1/samples/Sample_A50/C#_SBXPCSample/SBXPCSampleCSharp/frmAccessTz.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

using System.Runtime.InteropServices;

namespace SBXPCSampleCSharp
{
    public partial class frmAccessTz : Form
    {
        public frmAccessTz()
        {
            InitializeComponent();
        }

        AxSBXPCLib.AxSBXPC bpc;
        const int DBTIMEZONE_COUNT = 50;
        const int DBTIMESECTION_COUNT = 8;
        const int DBALLCOUNT = DBTIMEZONE_COUNT * DBTIMESECTION_COUNT;

        Byte[] timeZoneInfo;
        private void frmAccessTz_Load(object sender, EventArgs e)
        {
            bpc = (AxSBXPCLib.AxSBXPC) Application.OpenForms["frmMain"].Controls["SBXPC1"];
            timeZoneInfo = new Byte[DBALLCOUNT * 5];
            dtStart.ShowUpDown = true;
            dtEnd.ShowUpDown = true;
            TimeZoneInit();
            DrawTimeZone();
        }

        private void TimeZoneInit()
        {
            for(int i = 0; i < DBTIMEZONE_COUNT; i ++)
            {
                for(int j = 0; j < DBTIMESECTION_COUNT; j ++)
                {
                    timeZoneInfo[(i * DBTIMESECTION_COUNT + j) * 5] = 0;
                    timeZoneInfo[(i * DBTIMESECTION_COUNT + j) * 5 + 1] = 0;
                    timeZoneInfo[(i * DBTIMESECTION_COUNT + j) * 5 + 2] = 23;
                    timeZoneInfo[(i * DBTIMESECTION_COUNT + j) * 5 + 3] = 59;
                    timeZoneInfo[(i * DBTIMESECTION_COUNT + j) * 5 + 4] = 0;
                }

            }
        }

        private void DrawTimeZone()
        {
            string itemString = "";
            lstTimeZone.Items.Clear();
            int startHour, startMinute, endHour, endMinute, verifyMode;
            for (int i = 0; i < DBTIMEZONE_COUNT; i ++ )
            {
                for (int j = 0; j < DBTIMESECTION_COUNT; j ++ )
                {
                    startHour   = timeZoneInfo[(i * DBTIMESECTION_COUNT + j) * 5];
                    startMinute = timeZoneInfo[(i * DBTIMESECTION_COUNT + j) * 5 + 1];
                    endHour     = timeZoneInfo[(i * DBTIMESECTION_COUNT + j) * 5 + 2];
                    endMinute   = timeZoneInfo[(i * DBTIMESECTION_COUNT + j) * 5 + 3];
                    verifyMode = timeZoneInfo[(i * DBTIMESECTION_COUNT + j) * 5 + 4];

                    itemString  = "[Tz.]" + String.Format("{0:D2}-{1:D1} ", i, j);
                    itemString += "[S]" + String.Format("{0:D2}:{1:D2} ", startHour, startMinute);
                    itemString += "[E]" + String.Format("{0:D2}:{1:D2}", endHour, endMinute) + " ";
                    itemString += "[VM]" + cmbVerifyMode.Items[verifyMode];
                    lstTimeZone.Items.Add(itemString);
                }
            }
        }

        private void cmdRead_Click(object sender, EventArgs e)
        {
            Boolean bRet;
            int vErrorCode = 0;
            String strXML = "";

            lblMessage.Text = "Waiting...";
            Application.DoEvents();
            
            if (!bpc.EnableDevice(Program.gMachineNumber, 0))
            {
                lblMessage.Text = util.gstrNoDevice;
                return ;
            }

            GCHandle gh = GCHandle.Alloc(timeZoneInfo, GCHandleType.Pinned);
            IntPtr AddrOftimeZoneInfo = gh.AddrOfPinnedObject();
            int nAddr = AddrOftimeZoneInfo.ToInt32();
            
            util.MakeXMLRequestHeader(bpc, ref strXML, "GetTimezone");
            
            bRet = bpc.GeneralOperationXML(ref strXML);
            bpc.XML_ParseBinaryLong(ref strXML, "TimezoneBinary", ref nAddr, DBALLCOUNT * 5);
            if (bRet)
            {
                DrawTimeZone();
                lblMessage.Text = "Success!";
            }
            else
            {
                bpc.GetLastError(ref vErrorCode);
                lblMessage.Text = util.ErrorPrint(vErrorCode);
            }
            
            bpc.EnableDevice (Program.gMachineNumber, 1);
        }

        private void frmAccessTz_FormClosing(object sender, FormClosingEventArgs e)
        {
            Application.OpenForms["frmMain"].Visible = true;
        }

        private void cmdExit_Click(object sender, EventArgs e)
        {
            Close();
        }

        private void cmdWrite_Click(object sender, EventArgs e)
        {
            Boolean bRet;
            int vErrorCode = 0;
            String strXML = "";

            lblMessage.Text = "Waiting...";
            Application.DoEvents();

            if (!bpc.EnableDevice(Program.gMachineNumber, 0))
            {
                lblMessage.Text = util.gstrNoDevice;
                return;
            }

            GCHandle gh = GCHandle.Alloc(timeZoneInfo, GCHandleType.Pinned);
            IntPtr AddrOftimeZoneInfo = gh.AddrOfPinnedObject();
            int nAddr = AddrOftimeZoneInfo.ToInt32();

            util.MakeXMLRequestHeader(bpc, ref strXML, "SetTimezone");
            bpc.XML_AddBinaryLong(ref strXML, "TimezoneBinary", ref nAddr, DBALLCOUNT * 5);

            bRet = bpc.GeneralOperationXML(ref strXML);
            if (bRet)
                lblMessage.Text = "Success!";
            else
            {
                bpc.GetLastError(ref vErrorCode);
                lblMessage.Text = util.ErrorPrint(vErrorCode);
            }

            bpc.EnableDevice(Program.gMachineNumber, 1); 
        }

        private void lstTimeZone_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (lstTimeZone.SelectedIndex == -1) 
                return;

            int index = lstTimeZone.SelectedIndex;

            dtStart.Value = new DateTime(   2000, 1, 1,         // Don't care year/month/date
                                            timeZoneInfo[index * 5], 
                                            timeZoneInfo[index * 5 + 1], 
                                            0
                                        );
            dtEnd.Value = new DateTime(     2000, 1, 1,         // Don't care year/month/date
                                            timeZoneInfo[index * 5 + 2], 
                                            timeZoneInfo[index * 5 + 3], 
                                            0
                                      );
            cmbVerifyMode.SelectedIndex = timeZoneInfo[index * 5 + 4];
        }

        private void cmdUpdate_Click(object sender, EventArgs e)
        {
            if (lstTimeZone.SelectedIndex == -1)
                return;

            int index = lstTimeZone.SelectedIndex;

            timeZoneInfo[index * 5] = Convert.ToByte(dtStart.Value.Hour);
            timeZoneInfo[index * 5 + 1] = Convert.ToByte(dtStart.Value.Minute);
            timeZoneInfo[index * 5 + 2] = Convert.ToByte(dtEnd.Value.Hour);
            timeZoneInfo[index * 5 + 3] = Convert.ToByte(dtEnd.Value.Minute);
            timeZoneInfo[index * 5 + 4] = Convert.ToByte(cmbVerifyMode.SelectedIndex);
            DrawTimeZone();
        }
    }
}
